opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/exemplar/exemplar_reservoir.py

```diff
@@ -0,0 +1,263 @@
+        self.__reset()
+        return exemplar
+
+    def __reset(self) -> None:
+        self.__value = 0
+        self.__attributes = {}
+        self.__time_unix_nano = 0
+        self.__span_id = None
+        self.__trace_id = None
+        self.__offered = False
+
+
+class FixedSizeExemplarReservoirABC(ExemplarReservoir):
+    """Abstract class for a reservoir with fixed size."""
+
+    def __init__(self, size: int, **kwargs) -> None:
+        super().__init__(**kwargs)
+        self._size: int = size
+        self._reservoir_storage: list[ExemplarBucket] = [
+            ExemplarBucket() for _ in range(self._size)
+        ]
+
+    def maxSize(self) -> int:
+        """Reservoir maximal size"""
+        return self._size
+
+    def collect(self, point_attributes: Attributes) -> list[Exemplar]:
+        """Returns accumulated Exemplars and also resets the reservoir for the next
+        sampling period
+
+        Args:
+            point_attributes The attributes associated with metric point.
+
+        Returns:
+            a list of :class:`opentelemetry.sdk.metrics.exemplar.Exemplar`s. Returned
+            exemplars contain the attributes that were filtered out by the aggregator,
+            but recorded alongside the original measurement.
+        """
+        exemplars = filter(
+            lambda e: e is not None,
+            map(
+                lambda bucket: bucket.collect(point_attributes),
+                self._reservoir_storage,
+            ),
+        )
+        self._reset()
+        return [*exemplars]
+
+    def _reset(self) -> None:
+        """Reset the reservoir."""
+        pass
+
+
+class SimpleFixedSizeExemplarReservoir(FixedSizeExemplarReservoirABC):
+    """This reservoir uses an uniformly-weighted sampling algorithm based on the number
+    of samples the reservoir has seen so far to determine if the offered measurements
+    should be sampled.
+
+    Reference:
+        https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#simplefixedsizeexemplarreservoir
+    """
+
+    def __init__(self, size: int = 1, **kwargs) -> None:
+        super().__init__(size, **kwargs)
+        self._measurements_seen: int = 0
+
+    def _reset(self) -> None:
+        super()._reset()
+        self._measurements_seen = 0
+
+    def offer(
+        self,
+        value: Union[int, float],
+        time_unix_nano: int,
+        attributes: Attributes,
+        ctx: Context,
+    ) -> None:
+        """Offers a measurement to be sampled."""
+        index = self._find_bucket_index(value, time_unix_nano, attributes, ctx)
+        if index != -1:
+            self._reservoir_storage[index].offer(
+                value, time_unix_nano, attributes, ctx
+            )
+
+    def _find_bucket_index(
+        self,
+        value: Union[int, float],
+        time_unix_nano: int,
+        attributes: Attributes,
+        ctx: Context,
+    ) -> int:
+        self._measurements_seen += 1
```

Comment:
Should the increment to `measurements_seen` occur AFTER finding the index? Otherwise we always skip the first bucket?

Reply:
Actually not because [`randrange`](https://docs.python.org/3/library/random.html#random.randrange) exclude the stop value. Therefore if we don't increment first, we try to execute `randrange(0, 0)` resulting in an exception:

```sh
ValueError: empty range for randrange() (0, 0, 0)
```